### PR TITLE
doc: Add documentation about Pin control

### PR DESCRIPTION
**Pull request by umapraseeda**
Add documentation about Pin control

Signed-off-by: Uma Praseeda <uma.praseeda@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/ug_pinctrl.rst [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+.. contents::
+   :local:
+   :depth: 2
+
+Pin control, also referred to as ``pinctrl``, is a Zephyr RTOS API introduced as an effort to standardize how SoC peripheral pins are configured.
+This enables mapping peripheral signals to the required pins as well as configuring necessary pin properties.
+For example, using pin control, you can route ``UART0`` ``RX`` signal to pin ``P1.01`` and configure the pin pull-up resistor.
+
+Before pin control, the same functionality was available in a vendor specific way.
+For the previous example, before pin control, Nordic SoCs had to be configured in the following way:
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:15]:
```suggestion
Before pin control, the same functionality was available in a vendor specific way.
For the previous example, Nordic SoCs had to be configured in the following way:
```

(repeats)

File: doc/nrf/ug_pinctrl.rst [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+.. code-block:: dts
+
+    &uart0 {
+        ...
+        rx-pin = <33>;
+        rx-pull-up;
+        ...
+    };
+
+Pin control involves *states*, where each state encodes the pin configurations for a state of the peripheral.
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:26]:
```suggestion
Pin control works using a **state model**, where each state encodes the pin configurations for a state of the peripheral.
```

note: `state` repeats too much?

File: doc/nrf/ug_pinctrl.rst [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+Following are the standard states:
+
+* ``default`` - Peripheral in working mode.
+* ``sleep`` - Peripheral in sleep mode.
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:30]:
```suggestion
* ``sleep`` - Peripheral in sleep or low power mode.
```

File: doc/nrf/ug_pinctrl.rst [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+A similar approach is also used in other systems like Linux.
+For more details including nRF specific bits, see the next section.
+
+For more general information, see the :ref:`Zephyr documentation on pin controllers <pinctrl-guide>`.
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:35]:
```suggestion
For more general information, see the :ref:`Zephyr documentation on pin control <pinctrl-guide>`.
```

File: doc/nrf/ug_pinctrl.rst [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+   &pinctrl {
+           /* Add your modifications here */
+   };
+
+All device pin configurations must be placed in child nodes of the ``pinctrl`` node, as shown in the following example:
+
+.. code-block:: dts
+
+   &pinctrl {
+     /* configuration for uart0 device, default state */
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:54]:
```suggestion
      /* configuration for uart0 device, default state */
```

File: doc/nrf/ug_pinctrl.rst [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    };
+
+Migration of the devicetree files
+*********************************
+
+To migrate the devicetree files of any nRF-based board, you must complete the following steps:
+
+1. Create a :file:`board-pinctrl.dtsi` file, defining each peripheral state pin configuration.
+2. Include this file in the :file:`board.dts` file.
+3. Replace all ``-pin`` related properties with pin control states.
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:139]:
```suggestion
3. Replace all ``-pin`` related properties with pin control states.
4. Enable pin control by adding ``CONFIG_PINCTRL=y`` to the :file:`board.conf`
```

File: doc/nrf/ug_pinctrl.rst [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+    /* old board.dts */
+    &uart0 {
+        ...
+        tx-pin = <5>;
+        rx-pin = <33>;
+        rx-pull-up;
+        ...
+    };
+.. code-block:: dts
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:153]:
```suggestion
    };

.. code-block:: dts
```

File: doc/nrf/ug_pinctrl.rst [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        };
+        uart0_sleep: uart0_sleep {
+            group1 {
+                psels = <NRF_PSEL(UART_TX, 0, 5)>,
+                        <NRF_PSEL(UART_RX, 1, 1)>;
+                low-power-enable;
+            };
+        };
+    };
+.. code-block:: dts
```
Comment (gmarull) [inline doc/nrf/ug_pinctrl.rst:174]:
```suggestion
    };

.. code-block:: dts
```

**Review by gmarull — COMMENTED** [review]
a few minor nitpicks.

I think we also missed an important point. The old `-pin` method is still supported if `CONFIG_PINCTRL=n`, so boards will still be able to use the old method for a while (it will be deprecated in Zephyr 3.1 and removed in 3.3). We should mention that, in case some customers can't afford to do this migration immediately.

File: doc/nrf/ug_pinctrl.rst (hunk 1 above)
Comment (umapraseeda) [reply doc/nrf/ug_pinctrl.rst]: Fixed

File: doc/nrf/ug_pinctrl.rst (hunk 2 above)
Comment (umapraseeda) [reply doc/nrf/ug_pinctrl.rst]: Fixed

**Review by b-gent — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Very nice job

**Pull request merged**